runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp

```diff
@@ -131,19 +131,28 @@ void LightningSimulator::NamedOperation(const std::string &name, const std::vect
 
     // Update tape caching if required
     if (this->tape_recording) {
-        this->cache_manager.addOperation(name, params, dev_wires, inverse);
+        this->cache_manager.addOperation(name, params, dev_wires, inverse, {},
+                                         {/*controlled_wires*/}, {/*controlled_values*/});
     }
 }
 
 void LightningSimulator::MatrixOperation(const std::vector<std::complex<double>> &matrix,
                                          const std::vector<QubitIdType> &wires, bool inverse)
 {
+    RT_FAIL_IF(!wires.size(), "Invalid number of qubits");
```

Comment:
Wouldn't it be better to compare `1<<wires.size()` with matrix width here?

Reply:
This is checked in the Lightning C++ API. I didn't want to add a duplicated check here. what do you think? 

Also, it should be:
```suggestion
    RT_FAIL_IF(wires.empty(), "Invalid number of qubits");
```
I'll fix this in other places too.

Reply:
But then why do we do the emptiness check at all? Should we rely on the Lightning entirely instead?

Reply:
Hmm this is a great question. 
The thing with relying on devices to implement correctness checks is that the experience will not be uniform, and some devices may even completely skip any verification. The runtime provides us with a convenient place to perform runtime checks across the board, which I actually like.
Of course for lots of gates the performance overhead might not be insignificant, so ideally we should also have a user option to skip them altogether.

Reply:
@dime10 I agree with what you said, with the comment that we are discussing is the Lightning driver of the runtime, that is, this code is dedicated to work with a specific device. Relying on the checking procedures of this particular device - is a valid choice in my opinion.
But then again, do you think we should check the full compatibility (num of qubits matching the matrix size? is it the only condition btw?) or just the non-emptiness?

Reply:
Ah good point, this is about the lightning driver. In this case I'm not sure, I would say since the driver can be considered as the same project as the underlying library I probably wouldn't repeat the matrix check if it is already performed elsewhere.

Wires being empty is also an interesting check, one could even allow it as a no-op but I think so far we've chosen to disallow it?

Reply:
I would say it is a good practice to allow no-wires ops in future - I think it would handle a number of corner-cases gracefully for someone.
Right now I think it is not possible: for `qml.Identity(wires=[])` Pennylane reports `ValueError: Identity: wrong number of wires. At least one wire has to be given.`
For now I think removing the checks would be fine. Or at least we should change the message text to reflect the condition that was really checked.

Reply:
@grwlf agree with your points. I double-checked and the reason we had added checks for empty-wires was due to do the fact that Lightning C++ API used to raise an obscure assertion instead of a value-error. Let's remove these checks from here and let the Lightning C++ API capture these cases. Thanks for bringing this up! :raised_hands:

Reply:
I don't think that's good given that there were obscure errors in lightning. I agree that there is no need to duplicate an existing check, but if we remove the check from here it should be added to the runtime.

Reply:
yeah, but these constrains should be ideally checked in the backend devices. In the runtime, we should only check the validity of `wires` based on QubitManager.  Check here [56e8e6a](https://github.com/PennyLaneAI/catalyst/pull/502/commits/56e8e6a8cf7d1687b70a6bad616b8d827e80695a)

Reply:
Why should these things ideally be checked by backend devices? We don't necessarily have control over what backend devices do.

Reply:
> Why should these things ideally be checked by backend devices? 

Because, these sort of limitations are forced by the backend devices. In the runtime, however, checking the validity of wires can be only in the scope of QubitManager (that checks if the given set of wires are allocated and activated on the device).

> Of course for lots of gates the performance overhead might not be insignificant, so ideally we should also have a user option to skip them altogether.

Would you suggest using `assert` so that we can control it with `NODEBUG`? https://en.cppreference.com/w/cpp/error/assert

Reply:
> In the runtime, however, checking the validity of wires can be only in the scope of QubitManager (that checks if the given set of wires are allocated and activated on the device).

The check in question is whether the gate was called with any qubits at all, which could definitely be done by the runtime at the logical level.

> Because, these sort of limitations are forced by the backend devices.

So you would recommend not to disallow empty qubits for all programs, because this is a lightning specific restrictions and other backends might be fine with no qubits as a no-op? Ok that argument I could agree with.

Considering all of the above I would still not remove this check though. If the current device code is deficient in its error checking than the driver can make up for it, why not?